train.py

```diff
@@ -0,0 +1,219 @@
+import itertools
+import sys
+import torch
+import pickle
+from tqdm import tqdm
+from torch.autograd import Variable
+from utils import *
+from BiLSTM_CRF import BiLSTM_CRF
+import numpy as np
+
+pre_emb = "data/glove.6B.100d.txt"
+name = "test"
+mapping_file = "models/mapping.pkl"
+models_path = "models/"
+model_name = models_path + name
+tmp_model = model_name + ".tmp"
+
+
+def evaluating(model, datas, best_F):
+
+    prediction = []
+    save = True
+    new_F = 0.0
+    confusion_matrix = torch.zeros((len(tag_to_id) - 2, len(tag_to_id) - 2))
+    for data in datas:
+        ground_truth_id = data['tags']
+        words = data['str_words']
+        chars = data['chars']
+        caps = data['caps']
+
+        len_chars = [len(c) for c in chars]
+        max_char = max(len_chars)
+        mask_char = np.zeros((len(len_chars), max_char), dtype='int')
+        for i, c in enumerate(chars):
+            mask_char[i, :len_chars[i]] = c
+
+        mask_char = Variable(torch.LongTensor(mask_char))
+
+        dwords = Variable(torch.LongTensor(data['words']))
+        dcaps = Variable(torch.LongTensor(caps))
+
+        val, out = model(dwords, mask_char, dcaps)
+        predicted_id = out
+
+        for (word, true_id, pred_id) in zip(words, ground_truth_id, predicted_id):
+            line = ' '.join([word, id_to_tag[true_id], id_to_tag[pred_id]])
+            prediction.append(line)
+            confusion_matrix[true_id, pred_id] += 1
+        prediction.append('')
+
+    predf = eval_temp + '/pred.' + name
+    scoref = eval_temp + '/score.' + name
+
+    with open(predf, 'w') as f:
+        f.write('\n'.join(prediction))
+
+    os.system('%s < %s > %s' % (eval_script, predf, scoref))
+
+    eval_lines = [l.rstrip() for l in open(scoref, 'r', encoding='utf8')]
+
+    for i, line in enumerate(eval_lines):
+        # print(line)
+        if i == 1:
+            new_F = float(line.strip().split()[-1])
+            if new_F > best_F:
+                best_F = new_F
+                save = True
+                # print('the best F is ', new_F)
+
+    # print(("{: >2}{: >7}{: >7}%s{: >9}" % ("{: >7}" * confusion_matrix.size(0))).format(

```

Comment:
don not just leave commented code, ideally remove it or at least say why is it commented and why you leave it here.